src/Controller/EloquaAppCloudEndpointController.php

```diff
@@ -0,0 +1,428 @@
+    $this->logger->debug('Received execute service hook with payload @records', [
+      '@record' => print_r($records, TRUE),
+    ]);
+    $pluginReferences = $this->getEntityPlugins($eloqua_app_cloud_service);
+
+    // Iterate over the ServiceEntity plugins and then over the payload items.
+    foreach ($pluginReferences as $pluginReference) {
+      $id = $pluginReference->value;
+      // Get the plugin manager from the list of plugins (from the container).
+      $pluginMgr = $this->plugins[$id];
+      // Instantiate the referenced plugin.
+      $plugin = $pluginMgr->createInstance($id);
+      // Is this a sync or async (bulk) plugin? If the annotation is empty then assume that it is async.
+      if (!empty($plugin->respond()) && $plugin->respond() === 'synchronous') {
+        $this->logger->debug('The @plugin plugin requested a synchronous response.', ['@plugin' => $plugin->getPluginId()]);
+        // Merge all the responses into one array.
+        // TODO: Will this even work?
+        $response = $this->respondSynchronously($plugin, $records, $instanceId, $executionId, $query);
+        $responseHtml = $this->renderer->renderRoot($response);
+        $result = new CacheableResponse($responseHtml);
+        $result->addCacheableDependency($response);
+        return $result;
+      }
+      else {
+        $this->logger->debug('The @plugin plugin requested an asynchronous response.', ['@plugin' => $plugin->getPluginId()]);
+        $response = $this->respondAsynchronously($plugin, $records, $instanceId, $executionId, $query);
+        $json = new JsonResponse($response);
+        $json->setStatusCode(204);
+      }
+    }
+    return $json;
+  }
+
+  /**
+   * @param $plugin
+   * @param $records
+   * @param $instanceId
+   *
+   * @return \stdClass
+   */
+  protected
+  function respondSynchronously($plugin, $records, $instanceId, $executionId, $query) {
+    $response = new \stdClass();
+    // The response will be the same for all contacts, but we need one "record".
+    $response = $plugin->execute($instanceId, new \stdClass(), $query);
+    return $response;
+  }
+
+  /**
+   * @param $plugin
+   * @param $records
+   * @param $instanceId
+   * @param $executionId
+   *
+   * @return \stdClass
+   */
+  protected
+  function respondAsynchronously($plugin, $records, $instanceId, $executionId, $query) {
+    /**
+     * Get the appropriate queue for this plugin.
+     *
+     * @var QueueInterface $queue
+     */
+    $queue = $this->queueFactory->get($plugin->queueWorker());
+    // Put the records directly onto on the queue.
+    foreach ($records as $record) {
+      // Let the plugin manipulate the record as needed.
+      $plugin->execute($instanceId, $record, $query);
+    }
+    // @TODO Define a queueItem class?
+    $queueItem = new \stdClass();
+    // Pass the queue type to the worker to make it easy to requeue if there are more then 5000 records.
+    $queueItem->queueId = $plugin->queueWorker();
```

Comment:
- [ ] Nit, the name is a little confusing. Would have made more sense to be `$plugin->queueWorkerId`?

Reply:
Good call, I went with $queueItem->queueWorker = $plugin->queueWorker(); for max consistency.